Initial work on pjc build

andreypopp opened the pull request:
WIP

andreypopp committed:
Initial work on pjc build

andreypopp commented on .bin/esy.js:
```diff
@@ -134,7 +134,10 @@ var esyEnv = require('./esyEnv.js');
 var builtInCommands = {
-  "build": true,
+  "build": function(env, args) {
+    var build = require('./esyBuildCommand');
```
Lazy import so we don't degrade startup perf.

jordwalke replied:
I like how you think.

andreypopp commented on .bin/esyBuildCommand.js:
```diff
@@ -0,0 +1,225 @@
+      ? `"${collectTransitiveDependencies(envMap, info)
+            .map(dep => `$_install_tree/node_modules/${dep}/lib`)
+            .join(':')}"`
+        : null,
```
This is hackish, should generate `findlib.conf` as a part of `pic build` preparations instead and not put a burden on a build tool.

andreypopp replied:
Fixed. Now emitted `Makefile` has all info to generate `findlib.conf`.

andreypopp commented on .bin/esyBuildCommand.js:
```diff
@@ -0,0 +1,225 @@
+  if (!onlyMeta) {
+    pkgEnv = pkgEnv.concat([
+      {
+        name: 'FINDLIB_CONF',
```
Should be `OCAMLFIND_CONF`, should be fixed in pjc docs too.

andreypopp replied:
Fixed! Actually I get why @jordwalke wanted this to be in `esy` — any build system which uses `ocamlfind` is now going to be compat automatically!

andreypopp commented on tests/TestOne/PackageA/package.json:
```diff
@@ -4,7 +4,11 @@
   "dependencies": {
-    "PackageB": "../PackageB"
+    "PackageB": "../PackageB",
+    "buildtool": "../buildtool"
```
A simplest pjc-compat build tool is also modelled as being distributed as npm package.

jordwalke replied:
Sounds great!

andreypopp commented on tests/TestOne/PackageA/package.json:
```diff
@@ -4,7 +4,11 @@
+    "buildtool": "../buildtool"
+  },
+  "pjc": {
+    "build": "$buildtool__root/bin/buildtool"
```
We see `buildtool` as a binary package (it includes ready to run executable at package's root and don't require build).

jordwalke replied:
I have been putting things into `PATH` (you can do that by having a dependency export a globally visible environment variable to its dependers) but I might actually like this better - it's more explicit about which dependency a binary comes from!

andreypopp commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,83 @@
+#
+# This is a very simple pjc-aware ocamlfind-based build tool.
+
+export OCAMLFIND_CONF=$cur_target_dir/findlib.conf
```
Should be done by `pjc/esy build`

andreypopp commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,83 @@
+DEPENDS=""
+for p in $cur_depends; do
+  if [ "$p" != "buildtool" ]; then
+    DEPENDS="$DEPENDS $p"
```
We filter out `buildtool` itself here but `buildtool` should be modelled as `buildTimeDependency` instead.

andreypopp commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,83 @@
+cat <<EOF > $cur_target_dir/findlib.conf
+path = "$FINDLIB_PATH:$cur_depends_install_closed"
+destdir = "$cur_install"
+EOF
```
Should be done by pjc/esy build

andreypopp commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,83 @@
+  ocamlfind ocamlc \
+    -verbose \
+    -linkpkg $OCAMLFIND_ARGS \
+    $obj -o $target
```
This is where it fails right now... can't figure out why yet. Relevant trace:
```
+ ocamlfind ocamlc -verbose -linkpkg -package PackageB /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_build/node_modules/PackageA/bin/package_a_cmd.cmo -o /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_build/node_modules/PackageA/package_a_cmd
Effective set of compiler predicates: pkg_PackageC,pkg_PackageB,autolink,byte
+ ocamlc -verbose -o /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_build/node_modules/PackageA/package_a_cmd -I /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_install/node_modules/PackageC/lib/PackageC -I /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_install/node_modules/PackageB/lib/PackageB /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_install/node_modules/PackageC/lib/PackageC/PackageC.cma /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_install/node_modules/PackageB/lib/PackageB/PackageB.cma /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_build/node_modules/PackageA/bin/package_a_cmd.cmo
File "_none_", line 1:
Error: Error while linking /Users/andreypopp/Workspace/esy/tests/TestOne/PackageA/_build/node_modules/PackageA/bin/package_a_cmd.cmo:
Reference to undefined global `PackageB'
ocamlc returned with exit code 2
```

andreypopp replied:
Works, now, archive files were produced empty before.

andreypopp commented:
After cloning the repo:
```
npm install
cd tests/TestOne/PackageA
make build # tries to build PackageA --> PackageB --> PackageC
```

andreypopp committed:
Simplify test setup

Use process substituion not to generate temporary Makefile.

In the future we could make command `build-eject` which prints Makefile on
stdout and then command `build` will just do `make -f <(esy build-eject)`.

andreypopp committed:
Move findlib.conf generation to esy build command

andreypopp committed:
Simplify a little bit

andreypopp committed:
Fix buildtool to correctly produce archives

andreypopp committed:
Less verbose output

andreypopp committed:
Configure sandbox via ESY__SANDBOX

jordwalke commented on .bin/esyBuildCommand.js:
```diff
@@ -0,0 +1,323 @@
+}
+
+function getPkgEnv(envMap, pkg, current = false) {
+  let prefix = current ? 'cur' : (pkg.packageJson.name + '_');
```
I think this should be `'cur_'`

jordwalke commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,69 @@
+# Build
+
+DEPENDS=""
+for p in $cur_depends; do
```
I think this should be `$cur__depends` (double underscore to signify that it's part of the ecosystem - not any specific build script). (I commented elsewhere, where the corresponding fix would go).

jordwalke replied:
Same with `cur_target_dir` which should be `cur__target_dir`. The pjc docs were wrong so I'll update them now.

jordwalke commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,69 @@
+## Build modules
+
+for src in `find $cur_root/lib -name "*.ml"`; do
+  ocamlfind ocamlc -c $OCAMLFIND_ARGS $src -o $cur_target_dir/lib/`basename $src`
```
This would need to use `ocamldep` to make sure they're compiled in the right order, but this is just meant to be a dirt simple example so I wouldn't worry about it.

jordwalke commented on .bin/esyBuildCommand.js:
```diff
@@ -0,0 +1,323 @@
+    },
+    {
+      name: `${prefix}_install`,
+      value: `$_install_tree/node_modules/${info.name}/lib`,
```
Now, in the `pjc` docs, I've said that the following: `_install_tree`, `sandbox`, `_build_tree` should be prefixed with `pjc__` ( `pjc__install_tree`, `pjc__sandbox`, `pjc__build_tree`) which means we should make them be ( `esy__install_tree`, `esy__sandbox` (already done), `esy__build_tree`).

jordwalke replied:
I'm not sure if we should go with upper/lower case. Thoughts?

jordwalke commented on tests/TestOne/buildtool/bin/buildtool:
```diff
@@ -0,0 +1,69 @@
+
+echo "Install location: $cur_install"
+
+mkdir -p $cur_install
```
Same here (`cur__install`).
This is totally the right way to do this in general though. This is a general purpose build script that needs to work dynamically on the "cur" package - whichever that may be!

jordwalke commented on tests/TestOne/PackageA/Makefile:
```diff
@@ -0,0 +1,10 @@
+	sh ./test.sh
+
+clean:
+	rm -rf _build _install
```
I think we want these to somehow be `cur__target_dir` and `cur__install`. The challenge seems to be that these variables are set by by the make file spit out by `esy build` command above. Maybe we create another command `esy build-env` that creates a makefile that *only* has sets the environment variables for a package build, but doesn't actually build anything. Then the `clean` step could be:
```
@$(MAKE) -f <(../../../.bin/esy build-env) PackageA && rm -rf $cur__target_dir cur_install
```

jordwalke reviewed (approved):
Accepting/merging as is! We'll fix the small naming improvements next.

jordwalke commented on .bin/esyBuildCommand.js:
```diff
@@ -0,0 +1,323 @@
+      multiline: true,
+      name: `${info.name}__FINDLIB_CONF`,
+      value: `
+path = "$(shell ocamlfind printconf path):${findlibPath}"
```
One challenge is that this makes `ocamlfind` a mandatory dependency of `esy`. It's not bad if it's a requirement of the example build system, but I feel like we shouldn't need it for `esy` itself.

(Even then, it would be nice to not require `ocamlfind` for the example build system either - we can make a mock `ocamlfind` script).

andreypopp replied:
I think this is a temporary solution... ideally we would want to specify a dependency on a compiler — that will bring us a set of compiler libs alone, right?

jordwalke replied:
I think the solution here is that `esy` *tells* ocamlfind where it is going to store things, and then we never have to *ask* it, or even depend on it. We simply set the right environment variables.

Then when individual *packages* depend on `ocamlfind`, `esy` will have prepared all the install destinations, created all the config, and set all the variables to the right values so that everything can "just work".
